internal/controller/controllers/agent_controller.go

```diff
@@ -204,6 +209,132 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 	return r.updateStatus(ctx, log, agent, &h.Host, h.ClusterID, nil, false)
 }
 
+func findMatchingNode(agent *aiv1beta1.Agent, nodes []corev1.Node) *corev1.Node {
+	addresses := make([]string, 0)
+	for _, intf := range agent.Status.Inventory.Interfaces {
+		for _, ip := range append(intf.IPV4Addresses, intf.IPV6Addresses...) {
+			splitIp := strings.Split(ip, "/")
+			if len(splitIp) == 2 {
+				addresses = append(addresses, splitIp[0])
+			}
+		}
+	}
+	for i := range nodes {
+		n := &nodes[i]
+		if strings.EqualFold(n.Name, agent.Spec.Hostname) {
+			return n
+		}
+		for _, addr := range n.Status.Addresses {
+			if addr.Type == corev1.NodeInternalIP && funk.ContainsString(addresses, addr.Address) {
+				return n
+			}
+		}
+	}
+	return nil
+}
+
+func isNodeReady(node *corev1.Node) bool {
+	for _, c := range node.Status.Conditions {
+		if c.Status == corev1.ConditionTrue && c.Type == corev1.NodeReady {
+			return true
+		}
+	}
+	return false
+}
+
+func isNodeApproved(agent *aiv1beta1.Agent, nodes []corev1.Node) bool {
+	if agent.Status.DebugInfo.State == models.HostStatusAddedToExistingCluster &&
+		agent.Status.Progress.CurrentStage == models.HostStageRebooting {
+		node := findMatchingNode(agent, nodes)
+		if node != nil {
+			return isNodeReady(node)
+		}
+	}
+	return false
+}
+
+func (r *AgentReconciler) shouldApproveCSRs(agent *aiv1beta1.Agent, clients ClusterKubeClient) (bool, error) {
+
+	nodes, err := clients.ListNodes()
+	if err != nil {
+		return false, err
+	}
+	return !isNodeApproved(agent, nodes.Items), nil
+}
+func isCsrApproved(csr *certificatesv1.CertificateSigningRequest) bool {
+	for _, c := range csr.Status.Conditions {
+		if c.Type == certificatesv1.CertificateApproved {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *AgentReconciler) isAssociatedWithAgent(csr *certificatesv1.CertificateSigningRequest, agent *aiv1beta1.Agent) (bool, error) {
+	decodedCSR, _ := pem.Decode(csr.Spec.Request)
+	if decodedCSR == nil {
+		return false, errors.Errorf("Failed to decode request of CSR %s", csr.Name)
+	}
+	cr, err := x509.ParseCertificateRequest(decodedCSR.Bytes)
+	if err != nil {
+		return false, errors.Wrapf(err, "Failed to parse CSR %s", csr.Name)
+	}
+	return strings.EqualFold("system:node:"+agent.Spec.Hostname, cr.Subject.CommonName), nil
```

Comment:
this can be none not? if i didn't set it ?

Reply:
I think what @tsorya means is that the field could be empty if it's not set. Instead of using `agent.Spec` we should use the data in the Status. 

That said, the Status is set during the discovery phase, which may not be the right one. The way other parts of BMAC do it is to use the value from the Spec if set otherwise the one from the Status:


https://github.com/openshift/assisted-service/blob/1d327257c3cc54485c3cd09b517dc56907873967/internal/controller/controllers/bmh_agent_controller.go#L512-L517

Reply:
OK